feat: oracle version check

benesjan opened the pull request:
Fixes #14979

This PR implements a version compatibility system between Aztec.nr contracts and the PXE oracle interface to prevent ambiguous errors from version mismatches. The oracle version is expected to be manually updated but to ensure that this does not get forgotten the build throws an error if an oracle interface hash constant needs updating.

benesjan commented:
* **#16435** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16435?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16435?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on yarn-project/pxe/src/check_oracle_version.ts:
```diff
@@ -0,0 +1,36 @@
+ * changed and the ORACLE_VERSION constant needs to be incremented and the ORACLE_INTERFACE_HASH constant needs to be
+ * updated.
+ *
+ * TODO: The following only takes into consideration changes to the oracles defined in Oracle.ts and omits TXE oracles.
```
I don't know what is the future of the structure of TXE so I didn't implement this yet. A way how this could be done is to have the same check in TXE and have that be run on `TXE` class and checked against a `TXE_ORACLE_INTERFACE_HASH` constant. Why it didn't seem that great to do now is that the class doesn't contain only oracle functions. That class is very messy so think it would be preferable to first clean that up. If we don't want to deal with that in a foreseeable future than just doing the check on the messy class seems like the way to go.

WDYT?

nventuro replied:
Yeah once txe gets cleaned up we'll have a nice file with the full itnerface.

benesjan replied:
LMK if we have an issue for this (didn't find it in the [TXE Epic](https://github.com/AztecProtocol/aztec-packages/issues/9126)). If we don't have an issue would create a new one and link it in the code.

nventuro replied:
#16574 began introducing something like this, called `TXETypedOracle`. It's not in its final form however.

benesjan replied:
Created an [issue](https://github.com/AztecProtocol/aztec-packages/issues/16581) for this, added it to TXE epic since it's related and [linked it](https://github.com/AztecProtocol/aztec-packages/pull/16435/commits/54cc0c355070c3d177222739e41411f1cf66ab9c) in the code.

nventuro commented on noir-projects/aztec-nr/aztec/src/oracle/mod.nr:
```diff
@@ -22,6 +22,7 @@ pub mod logs;
 pub mod message_processing;
 pub mod notes;
 pub mod offchain_effect;
+pub mod oracle_version;
```
```suggestion
pub mod version;
```

nventuro commented on noir-projects/aztec-nr/aztec/src/test/helpers/test_environment.nr:
```diff
@@ -102,6 +106,7 @@ impl PrivateContextOptions {
 impl TestEnvironment {
     /// Creates a new `TestEnvironment`. This function should only be called once per test.
     pub unconstrained fn new() -> Self {
+        assert_oracle_version_matches();
```
We should have a test for this, probably using a noir oracle mock. We should also test calls to contract private and utility fns.

benesjan replied:
Testing this is problematic because the error is thrown in TS in the oracle handler so I am unable to test it with an oracle mock.

We could perform the assert in Noir but modifying "production" code for tests feels wrong. But agree that those tests are very good to have so unless we can think of some other option just doing the check in Noir is probably the best tradeoff.

Do you know if it would be somehow possible to modify the ORACLE_VERSION global constant in Noir in test setup? I assume not but just asking as I am not super sure.

 WDYT?

nventuro replied:
We can simply mock the oracle and then check that the oracle has been called, no need to have it fail. We already have tests were we check that it we call it with the wrong version the thing blows up.

benesjan replied:
You are absolutely right. Very smort Nico

Addressed in https://github.com/AztecProtocol/aztec-packages/pull/16435/commits/3086fa4f1dc7689c7f1012ef0e2f3bf726156dc1

Thanks

nventuro commented on yarn-project/pxe/src/check_oracle_version.ts:
```diff
@@ -0,0 +1,36 @@
+  if (oracleInterfaceHash !== ORACLE_INTERFACE_HASH) {
+    // This check exists only to notify you when you need to update the ORACLE_VERSION constant.
+    throw new Error(
+      `Oracle interface hash mismatch. Expected ${ORACLE_INTERFACE_HASH}, got ${oracleInterfaceHash}. Update ORACLE_INTERFACE_HASH and ORACLE_VERSION constants.`,
```
```suggestion
      `The TypedOracle interface has changed, which implies a breaking change in the aztec.nr/PXE oracle interface. Update Bump ORACLE_INTERFACE_HASH to ${oracleInterfaceHash} and bump ORACLE_VERSION in pxe/src/oracle_version.ts.`,
```

nventuro reviewed (commented):
Nice thinking with the bootstrap check

nventuro commented on noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr:
```diff
@@ -161,6 +161,7 @@ pub(crate) comptime fn transform_private(f: FunctionDefinition) {
         $storage_init
         $message_discovery_call
         $authorize_once_check
+        dep::aztec::oracle::version::assert_oracle_version_matches();
```
Shouldn't this go at the very top? Some oracles will have been called by this point.

benesjan replied:
Makes sense. Addressed in https://github.com/AztecProtocol/aztec-packages/pull/16435/commits/d1aa8dc3e61dca2d4e4838be2403c46fff012fd0

nventuro commented on noir-projects/aztec-nr/aztec/src/oracle/version.nr:
```diff
@@ -0,0 +1,33 @@
+mod test {
+    use super::assert_oracle_version_matches_oracle;
+
+    #[test(should_fail_with = "Oracle version mismatch. Expected version 1, got 318183437.")]
+    unconstrained fn oracle_version_mismatch() {
```
We should also have a test were we _don't_ blow up with the correct version.

benesjan replied:
Good point.

Addressed in https://github.com/AztecProtocol/aztec-packages/pull/16435/commits/2d20009f0c10108f56368d609439d8726f037e06

Thanks

nventuro commented on noir-projects/aztec-nr/aztec/src/oracle/version.nr:
```diff
@@ -0,0 +1,38 @@
+/// The ORACLE_VERSION constant is used to check that the oracle interface is in sync between PXE and Aztec.nr. We need
+/// to version the oracle interface to ensure that a developer gets a nice error message if he uses incompatible
```
```suggestion
/// to version the oracle interface to ensure that developers get a reasonable error message if they use incompatible
```

nventuro commented on noir-projects/aztec-nr/aztec/src/oracle/version.nr:
```diff
@@ -0,0 +1,38 @@
+/// versions of Aztec.nr and PXE. The TypeScript counterpart is in `oracle_version.ts`.
+///
+/// @dev Whenever a contract function or TXE test is run, the `utilityAssertOracleVersionMatches` is called and if
+/// there is a mismatch, an error is thrown.
```
```suggestion
/// @dev Whenever a contract function or Noir test is run, the `utilityAssertOracleVersionMatches` oracle
/// is called and if there is a mismatch, an error is thrown.
```

nventuro commented on noir-projects/aztec-nr/aztec/src/oracle/version.nr:
```diff
@@ -0,0 +1,38 @@
+    }
+}
+
+unconstrained fn internal_assert_oracle_version_matches() {
```
```suggestion
unconstrained fn assert_oracle_version_matches_oracle_wrapper() {
```

The naming convertion is:

```noir
fn foo() {
  foo_wrapper();
}

unconstrained foo_wrapper() {
  foo_oracle();
}

#[foo]
unconstrained foo_oracle() {
}
```

If `foo` is unconstrained then we just have `foo` call `foo_oracle` since we don't need the wrapper. The wrapper only exists because Noir does not let us call oracles from constrained fns. This is likely outdated, I'll create an issue.

nventuro replied:
https://github.com/noir-lang/noir/issues/9671

nventuro commented on noir-projects/aztec-nr/aztec/src/oracle/version.nr:
```diff
@@ -0,0 +1,38 @@
+pub global ORACLE_VERSION: Field = 1;
+
+/// Asserts that the version of the oracle matches the version expected by the contract.
+pub fn assert_oracle_version_matches() {
```
I keep becoming confused by this name - I think it's because 'matches' is strange when there are no params and we do no matching. Perhaps `assert_compatible_oracle_version`? `assert_oracle_interface_compatibility`?

benesjan replied:
Addressed in https://github.com/AztecProtocol/aztec-packages/pull/16435/commits/257e2ea7440a6ab0c8724497f3985d2b18a8b72c

nventuro commented on noir-projects/aztec-nr/aztec/src/test/helpers/test_environment/test.nr:
```diff
@@ -310,3 +311,19 @@ unconstrained fn private_public_and_utility_context_share_default_chain_id() {
+        mock.get_last_params::<Field>(),
+        ORACLE_VERSION,
+        "Oracle called with unexpected version",
+    );
```
There is no need to have these explanation strings on the assertiosn - they just make the test more verbose. `assert_eq(mock.times_called(), 1);` is perfectly clear.

nventuro replied:
I fact I would remove all comments from this test. The test name plus code is sufficiently clear.

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/16435/commits/be4dd5504d01074ec25004b0d00ff20fc45daab9

nventuro commented on noir-projects/noir-contracts/contracts/test/oracle_version_check_contract/src/main.nr:
```diff
@@ -0,0 +1,19 @@
+mod test;
```
You can just inline the `mod test`, no need to create lots of files. This being so simple it's easier to understand/review by having it all in one place - and the goal of tests is to be very clear re. what they're doing.

benesjan replied:
No longer relevant as the tests were moved to TS in https://github.com/AztecProtocol/aztec-packages/pull/16435/commits/b39404537ae8a028672eda680e55d08e9e552466

nventuro commented on noir-projects/noir-contracts/contracts/test/oracle_version_check_contract/src/test.nr:
```diff
@@ -0,0 +1,47 @@
+use crate::OracleVersionCheck;
+
+#[test]
+unconstrained fn oracle_version_is_checked_when_private_function_is_called() {
+    // We mock the oracle to be able to check that it gets called and that it gets called with the correct version
+    let mock = OracleMock::mock("utilityAssertOracleVersionMatches");
+
+    // Create test environment and deploy the contract
+    let mut env = TestEnvironment::new();
+    let caller = env.create_light_account();
+    let contract_address = env.deploy("OracleVersionCheck").without_initializer();
+
+    // Calling the private function should trigger the oracle version check
+    let _result =
+        env.call_private(caller, OracleVersionCheck::at(contract_address).private_function());
+
+    assert_eq(mock.times_called(), 1, "Expected oracle version check to be called exactly once");
+    assert_eq(
+        mock.get_last_params::<Field>(),
+        ORACLE_VERSION,
+        "Oracle called with unexpected version",
+    );
+}
```
This looks wrong. `TestEnvironment::new` calls the the compatibility check, and because you installed the mock _before_ the `new` call you're getting a hit to it. If the contract also called the oracle you'd see that the mock was called twice, which means the contract _is not_ calling it.

In general, you want to install the mocks right before they are used in order to prevent accidents like the one here. If you had the `OracleMock::mock` line right before when it's expecte dto be called (`env.call_private`) you would've noticed the mock is not being invoked.

---

The reason why the mock is not getting involved is because `env.call_private` does not execute Noir code in the current execution context. The javascript TXE code gets invoked, which loads a built artifact from a json file, creates a pxe instance and simulator and runs those. The current `nargo test` execution sees none of that - it is just waiting for the oracle invoked by `call_private` (`privateCallNewFlow`) to return. The fact that `call_private` results in a simulator running ACIR code and invoking oracles is almost accidental, it could've been python code or anything else - completely disconnected from this run.

Because of the above, I don't think we can test this property in this way. Perhaps we can instead have an e2e in which we `spy` the pxe_oracle_interface?

nventuro replied:
(the same comment from above re. comments in the test and assertion strings applies - rein in your cursor instance)

benesjan replied:
Very nice catch!

Ported these tests over to TS in https://github.com/AztecProtocol/aztec-packages/pull/16435/commits/b39404537ae8a028672eda680e55d08e9e552466

To be able to mock the oracle and spy on its calls I needed to move the check from `TypedOracle` to `ExecutionDataProvider`. Think it's completely fine as having the check directly in `TypedOracle` anyway seemed a bit incorrect as it didn't follow the general style of the class.

nventuro commented on yarn-project/pxe/src/check_oracle_version.ts:
```diff
@@ -0,0 +1,36 @@
+import { keccak256String } from '@aztec/foundation/crypto';
```
This should go in the `bin` directory.

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/16435/commits/55dbabcae0a8ad44c7d878a61bee7d52dc06fe7d

nventuro commented on yarn-project/txe/src/oracle/txe_oracle.ts:
```diff
@@ -381,6 +382,12 @@ export class TXE extends TXETypedOracle {
 
+  override utilityAssertOracleVersionMatches(version: number): void {
+    if (version !== ORACLE_VERSION) {
+      throw new Error(`Oracle version mismatch. Expected version ${ORACLE_VERSION}, got ${version}.`);
```
```suggestion
      throw new Error(`Oracle version mismatch. PXE is using version '${ORACLE_VERSION}', but got a request for '${version}'.`);
```

nventuro commented on yarn-project/txe/src/txe_service/txe_service.ts:
```diff
@@ -123,6 +123,11 @@ export class TXEService {
 
   // PXE oracles
 
+  utilityAssertOracleVersionMatches(version: ForeignCallSingle) {
+    this.executor.utilityAssertOracleVersionMatches(fromSingle(version).toNumber());
+    return toForeignCallResult([]);
```
```suggestion
  utilityAssertOracleVersionMatches(foreignVersion: ForeignCallSingle) {
    const version = fromSingle(version).toNumber();
    
    this.executor.utilityAssertOracleVersionMatches(version);
    
    return toForeignCallResult([]);
```

This is the code style introduced in https://github.com/AztecProtocol/aztec-packages/pull/16537

nventuro commented on yarn-project/yarn.lock:
```diff
@@ -1251,6 +1251,7 @@ __metadata:
     "@types/node": "npm:^22.15.17"
     jest: "npm:^30.0.0"
     jest-mock-extended: "npm:^4.0.0"
+    json-stringify-deterministic: "npm:1.0.12"
```
Do we need this? It seems to be a bit of an obscure unmaintained package - we should keep PXE's dependencies watertight.

benesjan replied:
We already use that library in our `stdlib` when computing a hash of a contract artifact. It's needed to compute the hash of the oracle interface.

We could replace this dependency with [canonicalize](https://www.npmjs.com/package/canonicalize) that seems to be better maintained. Can do that in a followup PR.

nventuro commented on noir-projects/aztec-nr/aztec/src/oracle/version.nr:
```diff
@@ -0,0 +1,38 @@
+pub global ORACLE_VERSION: Field = 1;
+
+/// Asserts that the version of the oracle is compatible with the version expected by the contract.
+pub fn assert_oracle_version_matches() {
```
You forgot to update the function name 😂 

This one should be called `assert_compatible_oracle_version`, then the wrapper would be `assert_compatible_oracle_version_oracle_wrapper`, and the oracle would be `assert_compatible_oracle_version_oracle`.

benesjan replied:
🤦  https://github.com/AztecProtocol/aztec-packages/pull/16435/commits/3d3e41a6d019f62695a64e8d205d11ace5200542

Thanks

benesjan committed:
feat: oracle version check

Fixes #14979

This PR implements a version compatibility system between Aztec.nr contracts and the PXE oracle interface to prevent ambiguous errors from version mismatches. The oracle version is expected to be manually updated but to ensure that this does not get forgotten the build throws an error if an oracle interface hash constant needs updating.
